### PR TITLE
Create "Getting started" tutorial.

### PR DESCRIPTION
**Pull request by m-renaud**
Create a Getting Started guide focused on newcomers to the Haskell ecosystem and Cabal. This page sits at the front of the Cabal user guide and walks through creating a new package, building and running it, and adding a dependency on a package from Hackage.

Rendered docs can be found at: https://m-renaud-haskell-cabal.readthedocs.io/en/mrenaud-docs-newcomer-tutorial/getting-started.html

Follow-up PRs:

- Link to `cabal (v2-)build` docs (when they are written)
- Write and link to section about `^>=` (there's no place that really explains it currently)
- More commentary about `:target` syntax (and write docs discussing target syntax to link to)
- Add section about "projects"

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by jneira** [comment]
Thanks for writing this, it will be a great improvement for beginners.
One note: it seems to me that, in general, it could be more friendly for windows users: 
* Installation method refer to an only *nix method (I know that it redirects to haskell platform but still)
* All console commands are posix, although some of them could be executed in cmd and/or powershell. 

Otoh not sure if the tutorial needs some changes to be executed.

I've got an error following the instructions cause the cabal file generated has not the adequate `cabal-version` with support for the caret operator (i've tried it with `3.0.0.0` and `3.2.0.0`):

```
PS D:\dev\ws\haskell\myapp> cabal-3.0.0.0 run :myapp
Errors encountered when parsing cabal file .\myapp.cabal:

myapp.cabal:24:46: error:
unexpected major bounded version syntax (caret, ^>=) used. To use this syntax the package need to specify at least 'cabal-version: 2.0'. Alternatively, if broader compatibility is important then use: >=1.0.0.0 && <1.1
expecting "." or "-"

   23 |   build-depends:       base >=4.12 && <4.13,
   24 |                        haskell-say ^>=1.0.0.0
      |                                              ^
```

I am getting errors with the target for `cabal-3.2.0.0` (is it a cabal bug??!?) and the lib `haskell-say` is not in hackage (i suppose it will be before pushing this to master):

```
PS D:\dev\ws\haskell\myapp> cabal-3.2.0.0 run :myapp
cabal-3.2.0.0.exe: Unrecognised target ':myapp'.
Expected a meta-target 'all', rather than 'myapp'.

PS D:\dev\ws\haskell\myapp> cabal-3.0.0.0 run :myapp # same for cabal-3.2.0.0 run all
Resolving dependencies...
cabal-3.0.0.0.exe: Could not resolve dependencies:
[__0] trying: myapp-0.1.0.0 (user goal)
[__1] unknown package: haskell-say (dependency of myapp)
[__1] fail (backjumping, conflict set: haskell-say, myapp)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: myapp, haskell-say
```

File: Cabal/doc/getting-started.rst [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+.. code-block:: console
+
+    $ mkdir myfirstapp && cd myfirstapp
+
+Once you have an empty directory we can initialize our package:
+
+.. code-block:: console
+
+    $ cabal init
```
[the remark this answers is not in the document]
Comment (phadej) [reply Cabal/doc/getting-started.rst:32]: In other words, we want something like

```
cabal init --cabal-version=2.4 --license=NONE
```

for now. The new defaults are not yet released.

File: Cabal/doc/getting-started.rst [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+This will generate the following files:
+
+.. code-block:: console
+
+    $ ls
+    CHANGELOG.md
```
Comment (phadej) [inline Cabal/doc/getting-started.rst:39]: The `Setup.hs` is still generated.

File: Cabal/doc/getting-started.rst [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+case, you can instruct ``cabal init`` to generate just a library (with
+``--lib``) or both a library and executable with (``--libandexe``); for the full
+set of options see ``cabal init --help``.
+
+``myfirstapp.cabal`` is Cabal's metadata file which describes your package and
+its dependencies. We'll be updating this file in a little bit when we add an
+external dependency to our package.
+
+.. note:: By default the name of the current directory will be used for the
+	  package name, you can change this with ``-p <package-name>``.
```
Comment (phadej) [inline Cabal/doc/getting-started.rst:55]: I think it's better to supply package name explicitly...

File: Cabal/doc/getting-started.rst [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+
+Running the application
+^^^^^^^^^^^^^^^^^^^^^^^
+
+As mentioned above, ``cabal init`` with no arguments generates a package with a
+single executable that prints ``"Hello, Haskell!"`` to the terminal. To run the
+executable enter the following command:
+
+``cabal run :myfirstapp``
```
Comment (phadej) [inline Cabal/doc/getting-started.rst:65]: I think it's better to supply package name explicitly...

... and then this command can be literal.

Now I get

```
cabal: Unknown target ':myfirstapp'.
The package getting-started has no component 'myfirstapp'.
```

error.

File: Cabal/doc/getting-started.rst [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+with ``cabal build``:
+
+``cabal build :myfirstapp``
+
+
+Adding dependencies
+^^^^^^^^^^^^^^^^^^^
+
+Next we'll add an external dependency to our application. The unit of
+shareability in the Haskell ecosystem is a `package`, the database of available
```
Comment (phadej) [inline Cabal/doc/getting-started.rst:90]: Cabal term is *repository*, or

> Hackage is the Haskell community's central package archive of open source software.

as Hackage itself puts it.

File: Cabal/doc/getting-started.rst [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+Haskell packages is called `Hackage <https://hackage.haskell.org/>`__.
+
+In our application, we'll use a package called `haskell-say
+<https://hackage.haskell.org/package/haskell-say>`__ to print text to the
```
Comment (phadej) [inline Cabal/doc/getting-started.rst:94]: I wonder if there are any "better" example package. If users go to see that package they'll find `package.yaml` and `stack.yaml` and no CI setup and ...

If we put something at the very front (*the first 3rd party package a new user may see*), we have to put something of the highest quality. I'm biased, so I won't suggest any specific package here.

**Comment by phadej** [comment]
As I commented, we should rather collect what we can simplify once 3.4 is released in separate issue, but write the guide with `3.2` in mind. Then this will be useful immediately.

**Comment by jneira** [comment]
Yeah, i guessed it would work with latest version. 
But i would try to make the guide retrocompatible as much as possible, without adding nothing about to be obsolete or introduce any complexity. I would use `cabal-3.0if there is no strong reasons to disallow it, but `3.2` would be nice too. The guide must note the required version of cabal in a visible place.

F.e. the caret operator is great but simply changing the example to be

```cabal
executable myfirstapp
       main-is: Main.hs
       build-depends:
           base >=4.11 && <4.12,
           haskell-say >=1.0.0.0 && < 1.1
```

You could remove the note about the caret operator and it would look more simple (uniformed?) and it would be compatible with `cabal-3.0` at once.
Hopefully users will be able to discover the caret operator later.

**Comment by phadej** [comment]
`cabal-install-3.0` supports caret operator as nicely. The `cabal-version: 2.4` is the important part in the generated `.cabal` file itself.

> Hopefully users will be able to discover the caret operator later.

They won't. It's safer to assume that they won't read beyond *Getting started* if things just work.

**Comment by jneira** [comment]
Mmm, yeah, cabal supports it even before, but `cabal-3.0 init` by default does not generate a `.cabal` file compatible, and without the caret operator `cabal init --cabal-version=2.4` would not be needed, but it is fine.

> They won't. It's safer to assume that they won't read beyond Getting started if things just work.

Fair enough.

File: Cabal/doc/getting-started.rst [hunk 7]
```diff
@@ -0,0 +1,159 @@
+Getting Started with Haskell and Cabal
+======================================
+
+Installing the Haskell toolchain
+--------------------------------
+
+To install the Haskell toolchain follow the `ghcup instructions
+<https://www.haskell.org/ghcup/>`__.
```
[the remark this answers is not in the document]
Comment (m-renaud) [reply Cabal/doc/getting-started.rst]: Done.

File: Cabal/doc/getting-started.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (m-renaud) [reply Cabal/doc/getting-started.rst]: Yeah, I had forgotten that there hasn't been a release since 2.4 was the default. Added the explicity `--cabal-version=2.4 --license=NONE -p myfirstapp`.

File: Cabal/doc/getting-started.rst (hunk 2 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst:41]: Added back, can remove after the next release.

File: Cabal/doc/getting-started.rst (hunk 3 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst]: Done.

File: Cabal/doc/getting-started.rst (hunk 4 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst:65]: Done. I guess I was assuming they were in a directory called `myfirstapp` in which case `cabal init` would create an executable named `myfirstapp` which you could then run with the `:<target>` syntax.

File: Cabal/doc/getting-started.rst (hunk 5 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst]: Inlined definition from Hackage.

File: Cabal/doc/getting-started.rst (hunk 6 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst:94]: I was looking for something very small, with zero dependencies, that did something fun, that works entirely in the terminal. We could easily add a test or two and a CI setup to that package if we want it to showcase everything (I can send a PR this week).

**Comment by m-renaud** [comment]
Hey @jneira, thanks for the feedback!

> Installation method refer to an only *nix method ...

Added a link to a Windows walkthrough for getting Haskell set up.


> All console commands are posix, although some of them could be executed in cmd and/or powershell.

tbh I haven't used Windows in years, is there anything specific I should change here?

> ...cabal-version with support for the caret operator...

Yeah, oops, I had forgotten that the change to switch `cabal init`s Cabal spec version to 2.4 isn't in a release yet, updated instructions to pass `--cabal-version=2.4` explicitly.

File: Cabal/doc/getting-started.rst (hunk 6 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst:94]: And I think as far as `package.yaml` and `stack.yaml` go, these are both very widely used tools in the community today (regardless of personal opinions of them), so exposing a newcomer to them isn't necessarily a bad idea.

File: Cabal/doc/getting-started.rst [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+   import HaskellSay (haskellSay)
+
+   main :: IO ()
+   main =
+     haskellSay "Hello, Haskell! You're using a function from another package!"
+
+``import HaskellSay (haskellSay)`` brings the ``haskellSay`` function from the
+module named ``HaskellSay`` into scope. The ``HaskellSay`` module is defined in
+the ``haskell-say`` packages that we added a dependency on above.
```
Comment (phadej) [inline Cabal/doc/getting-started.rst:131]: Related to your previous comment. I think it would be actually good to find some library which has few transitive dependencies.

File: Cabal/doc/getting-started.rst [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+Now you can build and re-run your code to see the new output:
+
+.. code-block:: console
+
+   $ cabal run
+       ________________________________________________________
```
Comment (phadej) [inline Cabal/doc/getting-started.rst:138]: there is output in between, we should use opportunity to explain what happens i.e. building and caching of dependencies. That's the whole point of using `cabal` and not just raw GHC!

File: Cabal/doc/getting-started.rst [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+          \    \  \    \  \-----------|
+           \    \  \    \  \          |
+            \    \  \    \  \---------|
+            /    /  /     \
+           /    /  /       \  \-------|
+          /    /  /    ^    \  \      |
+         /    /  /    / \    \  \ ----|
+        /    /  /    /   \    \
+       /____/  /____/     \____\
+
```
Comment (phadej) [inline Cabal/doc/getting-started.rst:159]: I think we should have a section *What next*. I don't know what to put there yet, I guess https://www.haskell.org/documentation/ is a good to have at the very least. Maybe we want highlight something from there, if it's good to know for beginner cabal users.

File: Cabal/doc/getting-started.rst (hunk 9 above)
Comment (phadej) [reply Cabal/doc/getting-started.rst:138]: ... but I don't have anything specific in mind yet, *fun* is tricky requirement. So this is good for now.

File: Cabal/doc/getting-started.rst [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+Initializing the application
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Start by creating a ``myfirstapp`` directory to hold the project files:
+
+.. code-block:: console
+
+    $ mkdir myfirstapp && cd myfirstapp
```
Comment (jneira) [inline Cabal/doc/getting-started.rst:28]: in windows powershell (present by default in last windows versions) could be
```
>mkdir myfirstapp ; cd myfirstapp
```
in raw dos
```
> md myfirstapp && cd myfirstapp
```

File: Cabal/doc/getting-started.rst [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+.. code-block:: console
+
+    $ cabal init --cabal-version=2.4 --license=NONE -p myfirstapp
+
+This will generate the following files:
+
+.. code-block:: console
+
+    $ ls
```
Comment (jneira) [inline Cabal/doc/getting-started.rst:40]: `ls` would work in powershell, in dos we have to use `dir` (the output will be different in any case but whatever)

Comment (phadej) [reply Cabal/doc/getting-started.rst:40]: It looks like that if we ask windows users to use *PowerShell* than there're only small tweaks needed? casing for `cmd` would make it all too messy and hard to follow.

File: Cabal/doc/getting-started.rst (hunk 11 above)
Comment (phadej) [reply Cabal/doc/getting-started.rst:28]:
```
mkdir myfirstapp ; cd myfirstapp
```

would work in unix shells to, but I think it's better to split into two separate commands

File: Cabal/doc/getting-started.rst (hunk 12 above)
Comment (jneira) [reply Cabal/doc/getting-started.rst:40]: yeah, powershell is fine nowadays imo, it is the default console in windows 10

File: Cabal/doc/getting-started.rst (hunk 10 above)
Comment (jneira) [reply Cabal/doc/getting-started.rst:159]: That is great, it could contain some short list of awesome things you can do with cabal, to motivate users to continue reading more advanced stuff, like:
```
* Create libraries, executables, test suites and benchmarks for your package
* Set up a project with multiple packages using `cabal.project`
* Upload your package to a common repository to share your code
* ....
```

File: Cabal/doc/getting-started.rst (hunk 8 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst:133]: I was thinking that in a (yet to be written) walkthrough on working with multiple packages in a project is where we could pull in more deps, wanted to keep this as simple as possible.

File: Cabal/doc/getting-started.rst (hunk 9 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst:140]: Yeah, lets discuss in a follow-up PR, would like to get a baseline checked in and then we can make tweaks from there.

File: Cabal/doc/getting-started.rst (hunk 10 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst:161]: Created a short blurb, lets bikeshed content on a follow-up PR.

File: Cabal/doc/getting-started.rst (hunk 11 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst]: Split into two separate commands.

File: Cabal/doc/getting-started.rst (hunk 12 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst:42]: Added some words about it working on unix shells and PowerShell if you're a Windows user.

File: Cabal/doc/getting-started.rst [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+         /    /  /    / \    \  \ ----|
+        /    /  /    /   \    \
+       /____/  /____/     \____\
+
+
+What Next?
+----------
+
+Now that you know how to set up a simple Haskell package using Cabal, check out
+some of thee resources on the Haskell website's `documentation page
```
Comment (jneira) [inline Cabal/doc/getting-started.rst:167]: Typo: the

File: Cabal/doc/getting-started.rst (hunk 8 above)
Comment (phadej) [reply Cabal/doc/getting-started.rst:133]: it's not about multiple packages. It's about showing that `cabal-install` does work for you. You *don't have to worry (as a beginner) if a dependency has own dependencies*, it **just works**  :)

File: Cabal/doc/getting-started.rst (hunk 13 above)
Comment (m-renaud) [reply Cabal/doc/getting-started.rst:167]: Oops, will fix in follow-up, good catch :)
